File: greeksTest.py
import os
import requests
import json
from dotenv import load_dotenv

load_dotenv()

API_KEY = os.getenv("MASSIVE_API_KEY")
BASE_URL = os.getenv("MASSIVE_BASE_URL")


def get_options_chain_snapshot(underlying_ticker, **kwargs):
    """
    Fetch options chain snapshot for a given underlying asset.

    Args:
        underlying_ticker: The underlying stock ticker (e.g., 'AAPL')
        **kwargs: Optional filters (strike_price, expiration_date, contract_type, limit)

    Returns:
        dict: API response containing options chain data with greeks
    """
    url = f"{BASE_URL}/snapshot/options/{underlying_ticker}"

    params = {"apiKey": API_KEY}
    params.update(kwargs)

    response = requests.get(url, params=params)
    response.raise_for_status()

    return response.json()


def print_options_chain(data):
    """Pretty print the options chain data with greeks."""
    print(f"Status: {data.get('status')}")
    print(f"Request ID: {data.get('request_id')}")
    print("-" * 80)

    results = data.get("results", [])
    print(f"Total contracts returned: {len(results)}\n")

    for option in results:
        details = option.get("details", {})
        greeks = option.get("greeks", {})
        last_quote = option.get("last_quote", {})
        last_trade = option.get("last_trade", {})

        print(f"Contract: {details.get('ticker', 'N/A')}")
        print(f"  Type: {details.get('contract_type', 'N/A').upper()}")
        print(f"  Strike: ${details.get('strike_price', 'N/A')}")
        print(f"  Expiration: {details.get('expiration_date', 'N/A')}")
        print(f"  Exercise Style: {details.get('exercise_style', 'N/A')}")

        print(f"  Greeks:")
        print(f"    Delta: {greeks.get('delta', 'N/A')}")
        print(f"    Gamma: {greeks.get('gamma', 'N/A')}")
        print(f"    Theta: {greeks.get('theta', 'N/A')}")
        print(f"    Vega:  {greeks.get('vega', 'N/A')}")

        print(f"  Implied Volatility: {option.get('implied_volatility', 'N/A')}")
        print(f"  Open Interest: {option.get('open_interest', 'N/A')}")

        if last_quote:
            print(f"  Last Quote: Bid ${last_quote.get('bid', 'N/A')} / Ask ${last_quote.get('ask', 'N/A')}")

        if last_trade:
            print(f"  Last Trade: ${last_trade.get('price', 'N/A')} ({last_trade.get('size', 'N/A')} contracts)")

        print()


if __name__ == "__main__":
    # Example: Get AAPL options chain snapshot
    ticker = "AAPL"

    print(f"Fetching options chain snapshot for {ticker}...")
    print("=" * 80)

    try:
        data = get_options_chain_snapshot(
            ticker,
            limit=10  # Limit to 10 contracts for readability
        )

        # Print formatted output
        print_options_chain(data)

        # Also print raw JSON for detailed analysis
        print("=" * 80)
        print("Raw JSON Response:")
        print("=" * 80)
        print(json.dumps(data, indent=2))

    except requests.exceptions.HTTPError as e:
        print(f"HTTP Error: {e}")
        print(f"Response: {e.response.text}")
    except Exception as e:
        print(f"Error: {e}")

File: src/bin/greeks_test.rs
use anyhow::Result;
use clap::Parser;
use periscope::{client::MassiveClient, Config, OptionsChainResponse};
use tracing::info;
use tracing_subscriber::EnvFilter;

#[derive(Parser, Debug)]
#[command(name = "greeks_test")]
#[command(about = "Fetch and display options chain data with Greeks")]
struct Args {
    /// Underlying ticker symbol
    #[arg(short, long, default_value = "AAPL")]
    ticker: String,

    /// Number of contracts to fetch
    #[arg(short, long, default_value = "10")]
    limit: i32,

    /// Output raw JSON
    #[arg(long)]
    json: bool,
}

fn print_options_chain(data: &OptionsChainResponse) {
    println!("Status: {}", data.status);
    println!("Request ID: {}", data.request_id);
    println!("{}", "-".repeat(80));
    println!("Total contracts returned: {}\n", data.results.len());

    for option in &data.results {
        if let Some(details) = &option.details {
            println!(
                "Contract: {}",
                details.ticker.as_deref().unwrap_or("N/A")
            );
            println!(
                "  Type: {}",
                details
                    .contract_type
                    .as_deref()
                    .unwrap_or("N/A")
                    .to_uppercase()
            );
            println!(
                "  Strike: ${}",
                details
                    .strike_price
                    .map(|p| p.to_string())
                    .unwrap_or_else(|| "N/A".to_string())
            );
            println!(
                "  Expiration: {}",
                details.expiration_date.as_deref().unwrap_or("N/A")
            );
            println!(
                "  Exercise Style: {}",
                details.exercise_style.as_deref().unwrap_or("N/A")
            );
        }

        println!("  Greeks:");
        if let Some(greeks) = &option.greeks {
            println!(
                "    Delta: {}",
                greeks
                    .delta
                    .map(|v| format!("{:.6}", v))
                    .unwrap_or_else(|| "N/A".to_string())
            );
            println!(
                "    Gamma: {}",
                greeks
                    .gamma
                    .map(|v| format!("{:.6}", v))
                    .unwrap_or_else(|| "N/A".to_string())
            );
            println!(
                "    Theta: {}",
                greeks
                    .theta
                    .map(|v| format!("{:.6}", v))
                    .unwrap_or_else(|| "N/A".to_string())
            );
            println!(
                "    Vega:  {}",
                greeks
                    .vega
                    .map(|v| format!("{:.6}", v))
                    .unwrap_or_else(|| "N/A".to_string())
            );
        } else {
            println!("    N/A");
        }

        println!(
            "  Implied Volatility: {}",
            option
                .implied_volatility
                .map(|v| format!("{:.4}", v))
                .unwrap_or_else(|| "N/A".to_string())
        );
        println!(
            "  Open Interest: {}",
            option
                .open_interest
                .map(|v| v.to_string())
                .unwrap_or_else(|| "N/A".to_string())
        );

        if let Some(quote) = &option.last_quote {
            println!(
                "  Last Quote: Bid ${} / Ask ${}",
                quote
                    .bid
                    .map(|v| format!("{:.2}", v))
                    .unwrap_or_else(|| "N/A".to_string()),
                quote
                    .ask
                    .map(|v| format!("{:.2}", v))
                    .unwrap_or_else(|| "N/A".to_string())
            );
        }

        if let Some(trade) = &option.last_trade {
            println!(
                "  Last Trade: ${} ({} contracts)",
                trade
                    .price
                    .map(|v| format!("{:.2}", v))
                    .unwrap_or_else(|| "N/A".to_string()),
                trade
                    .size
                    .map(|v| v.to_string())
                    .unwrap_or_else(|| "N/A".to_string())
            );
        }

        println!();
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    tracing_subscriber::fmt()
        .with_env_filter(EnvFilter::from_default_env())
        .init();

    let args = Args::parse();
    let config = Config::from_env()?;
    let client = MassiveClient::new(&config);

    info!("Fetching options chain for {}", args.ticker);
    println!("Fetching options chain snapshot for {}...", args.ticker);
    println!("{}", "=".repeat(80));

    let params = periscope::client::OptionsChainParams {
        limit: Some(args.limit),
        ..Default::default()
    };

    let data = client.get_options_chain(&args.ticker, Some(params)).await?;

    if args.json {
        println!("{}", serde_json::to_string_pretty(&data)?);
    } else {
        print_options_chain(&data);

        println!("{}", "=".repeat(80));
        println!("Raw JSON Response:");
        println!("{}", "=".repeat(80));
        println!("{}", serde_json::to_string_pretty(&data)?);
    }

    Ok(())
}
